internal/app/app.go

```diff
@@ -0,0 +1,153 @@
+
+// DryRun is a context key to indicate that we should not apply the proxy
+// settings. It is intended to be used in tests only.
+type DryRun struct{}
+
+type busObject struct {
+	ctx    context.Context
+	cancel context.CancelCauseFunc
+
+	mu sync.Mutex
+}
+
+type options struct {
+	timeout time.Duration
+	initCtx context.Context
+}
+type option func(*options)
+
+// App is the main application object.
+type App struct {
+	conn    *dbus.Conn
+	obj     *busObject
+	timeout time.Duration
+}
+
+// Apply is a function called via D-Bus to apply the system proxy settings.
+func (b *busObject) Apply(http, https, ftp, socks, no, mode string) *dbus.Error {
+	// Methods calls spin up separate goroutines, so ensure we don't run them in parallel
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	log.Infof("Called Apply(%s, %s, %s, %s, %s, %s)", http, https, ftp, socks, no, mode)
+
+	p, err := proxy.New(b.ctx, http, https, ftp, socks, no, mode)
+	// We need to cancel the context in a deferred function to get the final
+	// state of the error variable.
+	defer func() { b.cancel(err) }()
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	if b.ctx.Value(DryRun{}) == true {
+		return nil
+	}
+
+	err = p.Apply()
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	return nil
+}
+
+// New creates a new App object.
+func New(opts ...option) (a *App, err error) {
+	defer decorate.OnError(&err, "cannot initialize application")
+
+	// defaults
+	args := options{
+		timeout: 1 * time.Second,
+		initCtx: context.Background(),
+	}
+	// applied options
+	for _, o := range opts {
+		o(&args)
+	}
+
+	// Don’t call dbus.SystemBus which caches globally system dbus (issues in tests)
+	conn, err := dbus.ConnectSystemBus()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancelCause(args.initCtx)
+	obj := busObject{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	if err = conn.Export(&obj, dbusObjectPath, dbusInterface); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if err = conn.Export(introspect.NewIntrospectable(&introspect.Node{
+		Name: dbusObjectPath,
+		Interfaces: []introspect.Interface{
+			introspect.IntrospectData,
+			{
+				Name:    dbusInterface,
+				Methods: introspect.Methods(&obj),
+			},
+		},
+	}), dbusObjectPath, introspect.IntrospectData.Name); err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+
+	reply, err := conn.RequestName(dbusInterface, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		_ = conn.Close()
+		return nil, err
+	}
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		_ = conn.Close()
+		return nil, fmt.Errorf("D-Bus name already taken")
+	}
+
+	return &App{
+		conn:    conn,
+		obj:     &obj,
+		timeout: args.timeout,
+	}, nil
+}
+
+// Wait blocks until the context is cancelled or the timeout is reached,
+// returning an appropriate exit code and closing the D-Bus connection before
+// exiting.
+func (a *App) Wait() int {
+	defer a.conn.Close()
+
+	for {
```

Comment:
there is a timeout, but I think if you pass the context to New() on the App, then, you can cancel it whenever you want, then, you derive the dbus context from it and you are safe.

This way, you don’t need to have a timeout via optional parameters, and you can control from the caller (which is the main CLI there), can hook it up with ctrl+C signal and so on…